ext/MitsubaLoader/CMitsubaLoader.cpp

```diff
@@ -1197,14 +1213,49 @@ reg_t registers[REG_COUNT];
 void setCurrBSDFParams(in vec3 n, in vec3 L)
 {
 	vec3 campos = irr_glsl_SBasicViewParameters_GetEyePos(CamData.params.NormalMatAndEyePos);
-	irr_glsl_ViewSurfaceInteraction interaction = irr_glsl_calcFragmentShaderSurfaceInteraction(campos, WorldPos, n);
+	irr_glsl_IsotropicViewSurfaceInteraction interaction = irr_glsl_calcFragmentShaderSurfaceInteraction(campos, WorldPos, n);
 	irr_glsl_BSDFIsotropicParams isoparams = irr_glsl_calcBSDFIsotropicParams(interaction, L);
 	//TODO: T,B tangents
 	vec3 T = vec3(1.0,0.0,0.0);
 	vec3 B = vec3(0.0,0.0,1.0);
 	currBSDFParams = irr_glsl_calcBSDFAnisotropicParams(isoparams, T, B);
 }
 
+vec2 getAlpha(in uint op, in uint ndf, in instr_t instr, in bsdf_data_t data, in mat2 dUV)
+{
+	vec2 a;
+	if (op==OP_ROUGHDIFFUSE)
```

Comment:
work out `instr_getAlphaUTexPresence(instr)` before you enter the if (which could be changed to a switch with a `#define` or something so we can benchmark later)

Reply:
actually this whole line
```glsl
float a_ = textureOrF32(data.data[0].xy, instr_getAlphaUTexPresence(instr), dUV)*scale.x;
```
could be done outside the divergent flow control

Reply:
Actually I'll just go ahead and suggest the optimized version
```glsl
#ifdef MITSUBA_USE_SWTICH
        #define FIRST_CASE(X,C) switch(X) \
        { \
                case C:
        #define NEXT_CASE(X,C) break; \
                case C:
        #define DEFAULT_CASE break; \
                default:
        #define END_CASES break; \
        }
#else 
        #define FIRST_CASE(X,C) if (X==C)
        #define NEXT_CASE(X,C) else if (X==C)
        #define DEFAULT_CASE } else
        #define END_CASES
#endif


vec2 getAlpha(in uint op, in uint ndf, in instr_t instr, in bsdf_data_t data, in mat2 dUV)
{
        // if only scale could be on the same offset!
       vec2 scale;
        // would be more fun if roughness texture data was at the same offset
        // and if anisotropic second roughness could be on the same offset too for bsdf that use it
        uvec2 alphaUdata,alphaVdata;
	FIRST_CASE(op,OP_ROUGHDIFFUSE)
	{
		scale = decodeRGB19E7(data.data[1].zw).rg;
		alphaUdata = data.data[0].xy;
	}
	NEXT_CASE(op,OP_ROUGHDIELECTRIC)
	{
		scale = decodeRGB19E7(uvec2(data.data[1].w, data.data[2].x)).rg;
		alphaUdata = data.data[0].yz;
		alphaVdata = uvec2(data.data[0].w, data.data[1].x);
	}
	NEXT_CASE(op,OP_ROUGHCONDUCTOR)
	{
		scale = decodeRGB19E7(data.data[2].zw).rg;
		alphaUdata = data.data[0].xy;
		alphaVdata = data.data[0].zw;
	}
	NEXT_CASE(op,OP_ROUGHPLASTIC)
	{
		scale = decodeRGB19E7(uvec2(data.data[1].w, data.data[2].x)).rg;
		alphaUdata = data.data[0].yz;
	}
	NEXT_CASE(op,OP_ROUGHCOATING)
	{
		//later
	}
        END_CASES
        // external validation ensures we cant hit it on the transmissives
        vec2 a;
        a.x = textureOrF32(alphaUdata, instr_getAlphaUTexPresence(instr), dUV);
	if (ndf == NDF_AS)
                a.y = textureOrF32(alphaVdata, instr_getAlphaVTexPresence(instr), dUV);
        else
                a.y = a.x;
	return a*scale;
}

#undef FIRST_CASE
#undef NEXT_CASE
#undef DEFAULT_CASE 
#undef END_CASES
```

should be muuuch faster

Reply:
just reminding you to use the macro in future code so we can compare between `if-else` vs. `switch` performance

Reply:
?

Reply:
waiting on the switches

Reply:
you mean use this wrapper everywhere where there are if/else statements currently?

Reply:
only everywhere there is a something that could be a `switch` (i.e. different cases for a numerical parameter)

this is because I'm not sure about the performance characteristics of `switch` vs. `if-else` chain, so its useful to have a fast switch.